### ISSUE TITLE
Gnome Flatpak

### ISSUE THREAD

**Issue by ghost** [electron-userland/electron-builder#512]
Since everyone seems to be listing Linux distribution tools, I'll add gnome flatpak to the list as well: 

http://flatpak.org/

**Comment by develar** [comment]
Do you like flatpack or just report issue ;)?

**Comment by kunkinkan** [comment]
AppImage, Flatpak and Snappy Apps are the 3 modern solutions at the moment. That's why I created the issue :-)

**Comment by develar** [comment]
Moved to backlog to keep issue list clear.

**Comment by develar** [comment]
AppImage #504 supported.

**Comment by grafenhofer** [comment]
Flatpak is taking up in usage recently. It supports some features missing from appimage (support for automatic updates e.g. via proper integration in gnome software, KDE support seems to be coming,...). It could be time to reconsider the "backlog" tag.

**Comment by prcastro** [comment]
I agree. Flatpak is becoming the standard not only in GNOME, but within KDE and probably ElementaryOS environments. Seems also to be a non-centralized way to keep applications reproducible and sandboxed, and seems way more secure than the alternatives. The number of contributors is also bigger than AppImage and Snap. GNOME Apps, Spotify, LibreOffice, Skype and Telegram already have Flatpak builds.

**Comment by black-snow** [comment]
Anyone willing to tackle this? We have deb, rpm, AppImage and Snappy it would be a shame to not have flatpak as well.

**Comment by prcastro** [comment]
Maybe these links are of interest:

https://github.com/endlessm/electron-flatpak-base-app
http://blog.manuq.com.ar/posts/building-electron-apps-offline-for-flathub/

**Comment by develar** [comment]
Sorry, moved to backlog for now.

**Comment by lbssousa** [comment]
Flatpak is the last missing piece for electron-builder to become a trully complete tool for packaging Electron apps. Is it too hard to port electron-forge Flatpak implementation to electron-builder?

**Comment by laurent22** [comment]
A Flatpak version is also something that people frequently ask for in my application, so support for it would be very welcome.

**Comment by develar** [comment]
As I am going to switch to vanilla gnome Linux from macOS, flatpack will be supported this/next month.

**Comment by spiral2k** [comment]
any updates? thanks for the amazing work!

**Comment by goncalossilva** [comment]
@develar If you have yet to begin working on it, this is likely of interest: http://docs.flatpak.org/en/latest/electron.html

**Comment by julian-alarcon** [comment]
There is Snap support available (almost all the features from Flatpack, I like it more), but yeah, will be great to have also Flatpack

**Comment by fbruetting** [comment]
> There is Snap support available (almost all the features from Flatpack, I like it more), but yeah, will be great to have also Flatpack

Snaps do not provide Flatpak’s biggest feature – which is being decentralized. Furthermore, Snaps are just on Ubuntu and it’s derivatives really supported, whereas you will find Flatpak support on way more distributions.

**Comment by polarathene** [comment]
> As I am going to switch to vanilla gnome Linux from macOS, flatpack will be supported this/next month.

@develar Are you still using Gnome (or Linux for that matter)? It's understandable if you went back to macOS.

I'm not sure what's required to support Flatpak, did you get around to attempting it and run into any issues? Would [a reference project supporting Flatpak electron builds be of help?](https://github.com/endlessm/electron-installer-flatpak)

**Comment by TingPing** [comment]
> I'm not sure what's required to support Flatpak, did you get around to attempting it and run into any issues? Would a reference project supporting Flatpak electron builds be of help?

So there are two different ways to approach this:

- Using Flatpak to build an electron app in a sandbox. This is what the linked repo tries to do.
- Spitting out a .flatpak bundle (or maybe even commit into a repo) from a electron project. This is what this issue would be about.

There is sadly not a ton of overlap in those two tasks.

**Comment by snaggen** [comment]
As someone who have built flatpaks both as repositories, and as .flatpak bundles. And these are basically the same thing, since the .flatpak bundle is just an export of the repository.
//Mattias Eriksson



18 mars 2019 20:26 av notifications@github.com:

**Comment by TingPing** [comment]
@snaggen I'm not talking about the format, I'm talking about their usage.

`electron-builder` has no use for a project like Flathub which builds projects in a sandbox with the `flatpak-builder` tool.

It instead has use where an upstream project exports to their own infrastructure.

**Comment by polarathene** [comment]
> Spitting out a .flatpak bundle (or maybe even commit into a repo) from a electron project. This is what this issue would be about.

Ok, so would you say supporting `.flatpak` bundle output is difficult for electron-builder to support? Is there any particular issues that can be run into? If you have any idea how much work it'd be to add the support that'd be helpful too.

I'm not familiar with building flatpak apps, but perhaps if there is some guidance/advice in how to go about adding that support, I or someone else might be able to find the time to look into it ourselves and create a PR?

**Comment by TingPing** [comment]
> Ok, so would you say supporting .flatpak bundle output is difficult for electron-builder to support? Is there any particular issues that can be run into? If you have any idea how much work it'd be to add the support that'd be helpful too.

No I do not believe it would be that difficult. Just using the `flatpak build-*` commands to put things in the right place and export it should generally work fine.

See the man pages for `flatpak-build-init`, `flatpak-build`, `flatpak-build-finish`, `flatpak-build-export`.

Generally `.flatpak` bundles aren't very user friendly though, users lose out on simple updating. So it might be worth exposing the ability to export to a repository easily.

**Comment by klemy** [comment]
Hello, Flatpack would be truly awesome! [And there is a lot of more hype for it than Snapcraft.](https://trends.google.com/trends/explore?q=snapcraft,flatpack) I hope it will be soon included. Keep your good work!

**Comment by tidux** [comment]
@TingPing adding the ability to publish to a repository would get you "publish to Flathub.org" basically for free.  Flathub is the de facto standard distro agnostic place to host flatpaks.

**Comment by unsungNovelty** [comment]
There seems to be a lot of interest here for Flatpak. Hope this get's resolved soon.

**Comment by gitpaybot** [comment]
A bounty of *10 usd* was added to this issue. See task on [GitPay](https://gitpay.me/#/task/352)

**Comment by gitpaybot** [comment]
A bounty of *10 usd* was added to this issue. See task on [GitPay](https://gitpay.me/#/task/352)

**Comment by jawadalkassim** [comment]
any good news about flatpak support?

**Comment by polarathene** [comment]
@jawadalkassim if there is news about such, it'll be mentioned here.. Just subscribe to the issue like everyone else, optionally pitch in on the bounty on GitPay to show your interest/support (or :+1: the 1st message of the thread like others and hope someone works on it).

**Comment by patriziobruno** [comment]
Hi @polarathene, I tried to add a bounty to the task on GitPay, but didn't work. I got redirectd to https://gitpay.me/undefined

**Comment by alexanmtz** [comment]
@patriziobruno which payment method you tried?

**Comment by gitpaybot** [comment]
A bounty of *200 usd* was added to this issue. See task on [GitPay](https://gitpay.me/#/task/352)

**Comment by patriziobruno** [comment]
> @patriziobruno which payment method you tried?

@alexanmtz PayPal

**Comment by alexanmtz** [comment]
> > @patriziobruno which payment method you tried?
> 
> @alexanmtz PayPal

Thanks @patriziobruno, we just fixed this issue with Paypal payments, can you try again?

**Comment by dteare** [comment]
I tried to add a bounty for this issue to GitPay this morning and got this error:

![image](https://user-images.githubusercontent.com/110676/89346523-3234e100-d677-11ea-8f95-f7def9969df3.png)

I repeated the payment again before realizing that Brave's shields likely tickled the issue. The third payment went through.

Now there are 2 pending bounties that shouldn't be there. I tried to reach out to support but haven't heard back yet. Can those be cancelled?

**Comment by alexanmtz** [comment]
> I tried to add a bounty for this issue to GitPay as we'd love to use Flatpak for 1Password but I got this error:
> 
> ![image](https://user-images.githubusercontent.com/110676/89346523-3234e100-d677-11ea-8f95-f7def9969df3.png)
> 
> I repeated the payment again before realizing that Brave's shields likely tickled the issue. The third payment went through.
> 
> Now there are 2 pending bounties that shouldn't be there. I tried to reach out to support but haven't heard back yet. Can those be cancelled?
> 
> I'm sorry I didn't capture the console logs. I didn't think of it until later.

Thanks for reporting @dteare, they're cancelled, and we're investigating the cause, probably we have a bug for orders created without signup into the platform.

**Comment by dteare** [comment]
Thank you so much!!

In case it helps you reproduce the issue I was using Brave Version 1.11.104 Chromium: 84.0.4147.105 (Official Build) (64-bit).

Take care! 👋

++dave;

**Comment by kallisti5** [comment]
A lot of electron applications aren't discoverable due to the lack of Flatpak applications.   Flatpak issues opened to electron applications end up here.

Gnome Software no longer supports Snap's natively.  ArchLinux, Fedora, all don't show electron apps due to the lack of FlatPak support.

A bounty is growing on this one.  Isn't anyone interested in working on it?

**Comment by vladimiry** [comment]
@kallisti5, there is a lot of Electron-based apps published on https://flathub.org/ So it's already quite possible to publish the Electron-based app to the flatpack repository.

**Comment by gasinvein** [comment]
@vladimiry Almost none of that apps on Flathub are maintained by their authors. I won't be surprised if your ElectronMail is the only one.

**Comment by vladimiry** [comment]
> Almost none of that apps on Flathub are maintained by their authors

@gasinvein it doesn't mean that it's hard to maintain.

**Comment by gasinvein** [comment]
@vladimiry But still, as we can tell from the issues linked here, many developers see the lack of flatpak output format support in electron-builder as a blocker for publishing flatpaks.

**Comment by vladimiry** [comment]
@gasinvein, correct, and I hope it's more clear now that there is no blocker. More than, even if the electron-builder will be able to somehow prepare the standalone flatpack installation package it won't make much sense since the flatpak's way is maintaining the app/package-relative repository in https://github.com/flathub.

**Comment by polarathene** [comment]
Either way, the gist of it is that when devs are asked about Flatpak support, which they likely have no experience in, they may look into it and find this issue, have a quick read through and leave it as that as "not presently as easy/possible to support".

If electron-builder can help in someway, the present $420 USD bounty would be for that. Additionally probably some documentation or link to an existing guide that walks a developer through the process of supporting Flatpak (or utilizing Flathub). Once the process is less vague/effort to adopt, this issue could be closed.

**Comment by vladimiry** [comment]
> Additionally probably some documentation or link to an existing guide that walks a developer through the process of supporting Flatpak (or utilizing Flathub).

- Here is how to build and run the flatpak thing locally https://docs.flatpak.org/en/latest/first-build.html (especially useful on initial stage when you start trying flatpak stuff). This is a part of [5 minutes intro-like guide](https://docs.flatpak.org/en/latest/getting-started.html).
- How to submit the app to flathub https://github.com/flathub/flathub/wiki/App-Submission#how-to-submit-an-app
- Look [here](https://github.com/search?q=org%3Aflathub+org.electronjs.Electron2.BaseApp&type=code) for submitted Electron-based apps samples.

**Comment by Croydon** [comment]
> Additionally, even if the electron-builder will be able to somehow prepare the standalone flatpack installation package it won't make much sense since the flatpak's way is maintaining the app/package-relative repository in https://github.com/flathub.

Flatpak is decentralized. Flathub is probably the most popular Flatpak repository, but it is just one of them.

electron-builder has support for snap too, which is comparable in this regard.

**Comment by vladimiry** [comment]
> electron-builder has support for snap too, which is comparable in this regard.

Agree. Here is how to distribute flatpaks in offline/standalone mode https://docs.flatpak.org/en/latest/single-file-bundles.html (I have not tried this in action).

**Comment by lbssousa** [comment]
I would like to reiterate my comment 2 years ago: https://github.com/electron-userland/electron-builder/issues/512#issuecomment-398711781

Is it worth to implement electron-forge's Flatpak packaging capabilities to electron-builder?

**Comment by snaggen** [comment]
Creating a flatpak is not very hard. Creating it for an electron app turned out to be another story. A developer at Endless wrote a post a while ago describing the procedure , which included multiple custom scripts. I now found an flatpak electron installer by endless, which I assume is the result of all that work.
So, is it easy to create electron app, well it might be now if you use the workaround scripts people have been forced to create due to this issue still not being implemented. 

Is there some reason that flatpak support is not built? If building flatpaks is so simple as said in previous comments, it should be equally simple to officially support it?

**Comment by gasinvein** [comment]
@snaggen There are to different (kind of opposite) approaches to produce a flatpak of electron app: one involves running electron-builder inside the flatpak builder sandbox, and the other involves electron-builder itself calling flatpak build commands to create a bundle (just like it does for deb/rpm/snap). Current apps on Flathub use the former, while this issue is about the later.
See also [this](https://github.com/electron-userland/electron-builder/issues/512#issuecomment-474065787) post.

**Comment by vladimiry** [comment]
> while this issue is about the later

I think that's a rare case when that `later` approach is actually needed. It's a headache for complex apps since it's like turning the flatpak builder into another CI system.

**Comment by Qiplex** [comment]
Subscribed. It is not clear how to subscribe to a particular issue on GitBub. It starts working per repository. 
https://github.com/isaacs/github/issues/1844

**Comment by vially** [comment]
There is a preliminary pull-request at #5711 which should make it possible to build `*.flatpak` bundles. In case anyone is feeling adventurous and is willing to give it a try, the early feedback would be greatly appreciated and it would help in ironing out the potential issues faster.

However, as others have already mentioned in this thread, for Flatpak it would make a lot of sense to expose the functionality to publish the app to a repository (e.g.: Flathub) but that is **not** part of this pull-request. That's partly because sometimes it's better to do things incrementally, but in this case it's also because it wasn't entirely clear to me how that should work.

So if anyone has some input on this it I *might* take a stab at it in the future. In particular, it would help a lot if it was clear what the input and output of such a feature would be (while keeping in mind the constraints of the [Flathub App Submission](https://github.com/flathub/flathub/wiki/App-Submission) process).

**Comment by rathboma** [comment]
@vially You are amazing! I think everyone would be happy to see your PR merged, even if flathub submission requires another PR.

**Comment by mmaietta** [comment]
PR Merged. Closing this issue.
Any future improvements/features, such as Flathub, should probably be tracked in a new ticket.

@develar can you please run a full release cut? I think 22.10 has enough feature introductions that we should publish as the `latest` npm version.

**Comment by vially** [comment]
I think it may be worth pointing out that this should be considered experimental for now. 

I'm all for getting this out as soon as possible. But considering that I'm not very familiar with the `electron-builder` code base and since this hasn't received enough testing yet, it may be more prudent to **change the Flatpak target to be disabled by default** in the first release.

This way the early adopters can still manually enable it in their apps (which will help with ironing out the initial bugs) while keeping everyone else unaffected.

One big downside of Flatpak being enabled *by default* is the build-time dependency on [flatpak-builder](https://docs.flatpak.org/en/latest/flatpak-builder.html). This is a significant breaking change that needs to be documented properly because it has the potential to be *very disruptive* to the users of `electron-builder` who may or may not have `flatpak-builder` installed on their build machine (e.g.: dev machine or CI).

**Comment by mmaietta** [comment]
Certainly should be considered experimental and we'll need to get those docs working too, where we can label it as experimental.

I'll open a PR to remove it from the default target list. I missed that.

**Comment by gitpaybot** [comment]
hello ,

Here's is Gitpaybot with good news! You have a bounty to claim :-)

There's a bounty of $ 520 for a task you solved on Gitpay.

If you have a Pull Request merged, please claim on Gitpay (https://gitpay.me) on the issue:
https://gitpay.me/#/task/352

Please set the title of your pull request to *issue#352* and go in _Solve issue_ and send your Pull Request URL.

**Comment by johannesjo** [comment]
Thank you very much @vially !!

Does anybody know if there is already an issue that tracks a more convenient way to publish to flathub?  I don't know much about all the technical requirements for this, but [a lot of my users have been wanting this for a long time](https://github.com/johannesjo/super-productivity/issues/25).

**Comment by mmaietta** [comment]
There's currently no way to publish to flathub, but that would be a great integration to have. I'd be happy to advise if you are willing to contribute!
Here's a sample PR for Bitbucket publish and autoupdating https://github.com/electron-userland/electron-builder/pull/6228
Setting up a local dev environment is fairly simple as well: https://github.com/electron-userland/electron-builder/blob/master/CONTRIBUTING.md#to-setup-a-local-dev-environment

**Comment by gasinvein** [comment]
@johannesjo Making use of the recently-added flatpak support in electron-builder to publish on Flathub would require Flathub to accept binary uploads, what it currently doesn't allow. So, this is a more general issue than just about Electron. See flathub/flathub#2320 for a discussion about largely the same problem.
In the meantime, did you consider producing a flatpak as part of your CI pipeline without publishing it on Flathub?

**Comment by patriziobruno** [comment]
> @johannesjo Making use of the recently-added flatpak support in electron-builder to publish on Flathub would require Flathub to accept binary uploads, what it currently doesn't allow. So, this is a more general issue than just about Electron. See [flathub/flathub#2320](https://github.com/flathub/flathub/issues/2320) for a discussion about largely the same problem.
> In the meantime, did you consider producing a flatpak as part of your CI pipeline without publishing it on Flathub?

Flatpak was initially designed to be published using repositories, not bundle files. Bundle support bad been added, but it lacks all the advantages of having a repository: live updates, delta updates, and dependencies decoupling. Flathub is just one repository, if one doesn't want to use it, they can set one's up or publish to some other repository. Publishing flatpak bundles doesn't really work.

**Comment by johannesjo** [comment]
@gasinvein & @patriziobruno thank you very much for the explanations! Much appreciated! Probably makes sense to try to solve this via github actions in my case then. You don't happen to know a github actions example for an electron app by any chance?

**Comment by TheEvilSkeleton** [comment]
@johannesjo I'm going to assume that you will need to submit an MR to include the Electron BaseApp under here https://github.com/bilelmoussaoui/flatpak-github-actions/blob/47e4de41f760530ebad88296894877f2d2cab0aa/.github/workflows/docker.yml#L15, since flatpak-github-actions doesn't include it.

**Comment by gasinvein** [comment]
@patriziobruno That is true, bundles aren't very user-friendly, but I guess a flatpak bundle is better than no flatpak at all?

@johannesjo I'm not very familiar with electron, but I guess you'll need fairly recent electron-builder first? The version super-productivity uses doesn't have flatpak target support. Other than this, I guess it shouldn't need much changes other than adding `flatpak` to target list and installing flatpak (and probably flatpak-builder) on the builder machine.

**Comment by gasinvein** [comment]
@TheEvilSkeleton No, this action is for a different workflow, when electron-builder is running inside flatpak-builder sandbox, see [this](https://github.com/electron-userland/electron-builder/issues/512#issuecomment-716523117) comment.

**Comment by TheEvilSkeleton** [comment]
Ah, my bad.

**Comment by patriziobruno** [comment]
> @patriziobruno That is true, bundles aren't very user-friendly, but I guess a flatpak bundle is better than no flatpak at all?
> 
> @johannesjo I'm not very familiar with electron, but I guess you'll need fairly recent electron-builder first? The version super-productivity uses doesn't have flatpak target support. Other than this, I guess it shouldn't need much changes other than adding `flatpak` to target list and installing flatpak (and probably flatpak-builder) on the builder machine.
> 
> @TheEvilSkeleton No, this action is for a different workflow, when electron-builder is running inside flatpak-builder sandbox, see [this](https://github.com/electron-userland/electron-builder/issues/512#issuecomment-716523117) comment.

@gasinvein it's better to have flatpak bundles as a stepping stone to a flatpak repository. The discussion here and the resulting flatpak bundle implementation are priceless, but I hope they're not the end goal.

**Comment by vially** [comment]
There's been [some discussion](https://news.ycombinator.com/item?id=29523510) lately about the [sustainability of open-source](https://blog.filippo.io/professional-maintainers/) which reminded me of this issue (because it has a [bounty attached to it](https://gitpay.me/#/task/352/gnome-flatpak)). 

I've been considering claiming that bounty for a while now, but since the current Flatpak bundle support (#5711) is, at best, a partial solution, I think it might be better to transfer the bounty (either fully or partially) to a new issue that tracks "publishing Flatpaks to a repository" instead.

Therefore, my suggestion for everyone who contributed to the bounty (@ignapk, @dteare, @simonxciv and @patriziobruno) would be to decide whether they want to:

- a) transfer it to a different issue that better tracks their use-case/feature request
You might have to check what is the GitPay policy on these things or if they are even possible (/cc @alexanmtz).

- b) keep the bounty as is (in which case I'll claim it as part of this issue)


So just to make it absolutely clear, I'm still planning to claim the bounty attached to *this* issue, but **only after everyone who contributed to the bounty has had some time to decide what they want to do about it**.

Of course, everyone should feel free to suggest alternative solutions for this.

I would like to mention that, at least for me, it's a lot more important to get some closure and clarification on the scope of this issue than the bounty itself. If that means, for example, reopening this issue and defining its scope as "Publish Flatpaks to Flathub", that's a perfectly reasonable approach for me.


### Offtopic discussion about bug bounty platforms

This is slightly offtopic, but I think these bug bounty platforms work great when both the scope and the solution are somewhat clear. However, there are various challenges involved when:
- the scope is not clear enough 
- different sponsors have different (or even conflicting) requirements
- the solution is incomplete and/or implemented by multiple contributors
- etc

These are tough problems to solve and I'm not sure how they can be solved. But I think it's important to solve them if we want the bug bounty programs to succeed.

**Comment by puyoxyz** [comment]
so... is there an issue i can follow for flathub support ?

**Comment by mmaietta** [comment]
You can create one to encourage someone from the community to build support for a Flathub Publisher. I'm happy to review any code and provide examples of previously written publishers

**Comment by cbartondock** [comment]
Having some docs would really improve the situation. I feel like at the moment it's unclear if this works for all electron applications, or just some of them. It's also unclear to me which flat-pak options get exposed (all of module's sub-options for example?).

**Comment by Danik1601** [comment]
@cbartondock, someone [made](https://github.com/electron-userland/electron-builder/pull/6893) the documentations which was added in the [latest pre-release](https://github.com/electron-userland/electron-builder/releases/tag/v23.1.0)

**Comment by soredake** [comment]
https://github.com/electron-userland/electron-builder/issues/7154
